File: nodejs/orders/order_server.js
var grpc = require('@grpc/grpc-js');
var dapr = require("@dapr/dapr");
var messages = require('./pb/order_pb');
var services = require('./pb/order_grpc_pb');

const serverHost = "127.0.0.1";
const serverPort = "50051";
const client = new dapr.DaprClient({ serverHost, serverPort, communicationProtocol: dapr.CommunicationProtocolEnum.GRPC });
const ordersTable = "Customers.OrderTable"

async function placeOrder(call, callback) {
  let orderId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
  let orderDetail = call.request.getDesc();
  let custId = orderDetail.getCustomerid()
  let itemCost = orderDetail.getItemcost();
  let itemDesc = orderDetail.getItemdescription();

  console.log('placeOrder: start: id:' + orderId + ' order{customerId:' + custId + ', item:' + itemDesc + ', cost:' + itemCost + '}');

  const orderTs = Date.now() * 1e6
  const orderData = {
    key: orderId.toString(),
    value: {
      orderId: orderId.toString(),
      desc: {
        customerId: custId,
        itemCost: itemCost,
        itemDescription: itemDesc,
        orderTs: orderTs
      }
    }
  };
  const metadata = { };

  try {
    await client.state.save(ordersTable, [orderData], metadata);
    let reply = new messages.PlaceOrderReply();
    reply.setOrderid(orderId);
    reply.setTimestampinnanos(orderTs);
    callback(null, reply);
  } catch(err) {
    console.error('Error saving order:', err);
    callback(err);
  }
}

async function getOrder(call, callback) {
  console.log('getOrder: start: id:' + call.request.getOrderid());

  try {
    orderValue = await client.state.get(ordersTable, call.request.getOrderid().toString())
    if (orderValue && orderValue.desc) {
      let orderDetails = new messages.OrderDetails();
      orderDetails.setCustomerid(orderValue.desc.customerId);
      orderDetails.setItemcost(orderValue.desc.itemCost);
      orderDetails.setItemdescription(orderValue.desc.itemDescription);
      let order = new messages.Order();
      order.setTimestampinnanos(orderValue.desc.orderTs);
      order.setDesc(orderDetails);
      let reply = new messages.GetOrderReply();
      reply.setOrder(order);
      console.log('getOrder: complete: id:' + call.request.getOrderid());
      callback(null, reply);
    } else {
      console.error('getOrder: could not find order');
      callback(new Error('Order not found'), null);
    }
  } catch(err) {
    console.error('Error retrieving order:', err);
    callback(err);
  }
}

function main() {
  var server = new grpc.Server();
  server.addService(services.MyOrderServiceService, {placeOrder: placeOrder, getOrder: getOrder});
  server.bindAsync('127.0.0.1:26001', grpc.ServerCredentials.createInsecure(), (err, port) => {
    if (err != null) {
      return console.error(err);
    }
    console.log(`gRPC listening on ${port}`)
  });
}

main();

File: python/orders/order_server.py
#!/usr/bin/python3

import asyncio
import logging
import time
import random
import grpc

from dapr.clients import DaprClient
from dapr.clients.grpc._state import StateItem
from google.protobuf.json_format import Parse
from google.protobuf.json_format import MessageToJson

import pb.order_pb2
import pb.order_pb2_grpc

class Order(pb.order_pb2_grpc.MyOrderServiceServicer):
    daprClient = None
    # table name is of the format <database>.<table> where database and
    # table are what you've defined in your Bopmatic.yaml
    ordersTable = "Orders.prod.Customers.OrderTable"

    def __init__(self, daprClientIn):
        self.daprClient = daprClientIn

    async def PlaceOrder(
            self, request: pb.order_pb2.PlaceOrderRequest,
            context: grpc.aio.ServicerContext) -> pb.order_pb2.PlaceOrderReply:

        myOrderId = random.getrandbits(64)
        logging.info("PlaceOrder: start: id:%s order{customerId:%s, item:%s, cost:%s}",	myOrderId, request.desc.customerId, request.desc.itemDescription, request.desc.itemCost)
        order = pb.order_pb2.Order(desc = request.desc)
        order.timestampInNanos = time.time_ns()
        orderEncoded = MessageToJson(order)
        self.daprClient.save_state(store_name=self.ordersTable, key=format(myOrderId, 'x'), value=orderEncoded)
        logging.info("PlaceOrder: complete: id:%s", myOrderId)
        return pb.order_pb2.PlaceOrderReply(orderId=myOrderId, timestampInNanos=order.timestampInNanos)

    async def GetOrder(
            self, request: pb.order_pb2.GetOrderRequest,
            context: grpc.aio.ServicerContext) -> pb.order_pb2.GetOrderReply:

        logging.info("GetOrder: start: id:%s", request.orderId)
        state = self.daprClient.get_state(store_name=self.ordersTable, key=format(request.orderId, 'x'))
        myOrder = Parse(state.data, pb.order_pb2.Order())
        logging.info("GetOrder: complete: id:%s", request.orderId)
        return pb.order_pb2.GetOrderReply(order=myOrder)

async def serve() -> None:
    server = grpc.aio.server()
    random.seed(time.time_ns())
    try:
        with DaprClient() as myDaprClient:
            pb.order_pb2_grpc.add_MyOrderServiceServicer_to_server(Order(myDaprClient), server)
            listen_addr = '[::]:26001'
            server.add_insecure_port(listen_addr)
            logging.info("Starting server on %s", listen_addr)
            await server.start()
            await server.wait_for_termination()
    except Exception as e:
        print("Failed to open dapr client: ", e)

if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)
    asyncio.run(serve())
